src/Microsoft.ML.Transforms/Expression/BuiltinFunctions.cs

```diff
@@ -0,0 +1,1095 @@
+            Contracts.Assert(b >= 3);
+            if (a > 1290)
+                throw Contracts.Except($"Base must be at most 1290 when raising to the power of {b}");
+
+            // REVIEW: Should we use a checked context and exception catching like I8 does?
+            ulong u = (ulong)(uint)a;
+            ulong result = 1;
+            for (; ; )
+            {
+                if ((b & 1) != 0 && (result *= u) > I4.MaxValue)
+                    throw Contracts.Except("Overflow");
+                b >>= 1;
+                if (b == 0)
+                    break;
+                if ((u *= u) > I4.MaxValue)
+                    throw Contracts.Except("Overflow");
+            }
+            Contracts.Assert(result <= I4.MaxValue);
+
+            var res = (I4)result;
+            if (neg)
+                res = -res;
+            return res;
+        }
+
+        /// <summary>
+        /// Raise a to the b power. Special cases:
+        /// * 1^NA => 1
+        /// * NA^0 => 1
+        /// </summary>
+        public static I8 Pow(I8 a, I8 b)
+        {
+            if (a == 1)
+                return 1;
+            switch (b)
+            {
+                case 0:
+                    return 1;
+                case 1:
+                    return a;
+                case 2:
+                    return a * a;
+            }
+            if (a == -1)
+                return (b & 1) == 0 ? 1 : -1;
+            if (b < 0)
+                throw Contracts.Except("Cannot raise an integer to a negative power");
+
+            // Since the abs of the base is at least two, the exponent must be less than 63.
+            if (b >= 63)
+                throw Contracts.Except("Cannot raise an integer to a power greater than 62");
+
+            if (a == 0)
+            {
+                if (b == 0)
+                    return 1;
+                return 0;
+            }
+
+            bool neg = false;
+            if (a < 0)
+            {
+                a = -a;
+                neg = (b & 1) != 0;
+            }
+            Contracts.Assert(a >= 2);
+
+            // Since the exponent is at least three, the base must be < 2^21.
+            Contracts.Assert(b >= 3);
+            if (a >= (1L << 21))
+                throw Contracts.Except($"Base must be less than 2^21 when raising to the power of {b}");
+
+            long res = 1;
+            long x = a;
+            // REVIEW: Is the catch too slow in the overflow case?
+            try
+            {
+                checked
+                {
+                    for (; ; )
+                    {
+                        if ((b & 1) != 0)
+                            res *= x;
+                        b >>= 1;
+                        if (b == 0)
+                            break;
+                        x *= x;
+                    }
+                }
+            }
+            catch (OverflowException)
+            {
+                throw Contracts.Except("Overflow");
```

Comment:
Would we be better with [saturation arithmetic](https://en.wikipedia.org/wiki/Saturation_arithmetic) instead of throwing on overflow? Might be more of a PM-ish question of: is it better for a ML package to throw on semi-bad input or should we just do our best? I'm worried that this will throw in production when a model receives an unexpectedly large input (ML data is dirty).
```suggestion
                res = Int64.MaxValue;
```

Reply:
Resolved in follow-up PR: https://github.com/dotnet/machinelearning/pull/4614/files#r362319173

#Resolved